specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+
+ExperimentContainer
+    An ExperimentContainer is a QCoDeS object that stores all information about an experiment.
+    This includes items such as the equipment on which the experiment was run, the configuration of the equipment, graphs and other analytical output, and arbitrary notes, as well as the DataSet that holds the results of the experiment.
+
+Requirements
+============
+
+The DataSet class should meet the following requirements:
+
+Basics
+---------
+
+#. A DataSet can store data of (reasonably) arbitrary types.
+#. A completed DataSet should be immutable; neither its metadata nor its results may be modified.
+
+Creation
+------------
+
+#. It should be possible to create a DataSet without knowing the final item count of the various values it stores. 
+   In particular, the number of loop iterations for a sweep should not be required to create the DataSet.
+#. The list of parameters in each result to be stored in a DataSet should be specified at creation time.
+   This includes the name, role (set-point or output), and type of each parameter.
+   Parameters may be marked as optional, in which case they are not required for each result.
+#. It should be possible to define a result parameter that is independent of any QCoDeSParameter or Instrument.
+#. A QCoDeS Parameter should provide sufficient information to define a result parameter.
+#. A DataSet should allow storage of relatively arbitrary metadata describing the run that generated the results and the parameters included in the results.
+    
+Writing
+----------
+
+#. It should be possible to add a single result or a sequence of results to an in-progress DataSet.
+#. A DataSet should maintain the order in which results were added.
+#. An in-progress DataSet may be marked as completed.
+
+Access
+---------
+
+#. Values in a DataSet should be easily accessible for plotting and analysis, even while the DataSet is in progress.
+   In particular, it should be possible to retrieve results in a NumPy-compatible form.
+#. It should be possible to define a cursor that specifies a location in a specific value set in a DataSet.
+   It should be possible to get a cursor that specifies the current end of the DataSet when the DataSet is "in progress".
+   It should be possible to read "new data" in a DataSet; that is, to read everything after a cursor.
+#. It should be possible to subscribe to change notifications from a DataSet.
+   It is acceptable if such subscriptions must be in-process until QCoDeS multiprocessing is redone.
+   Change notifications should include the results that were added to the DataSet that triggered the notification.
+
+Storage and Persistence
+-----------------------
+
+#. A DataSet object should allow writing to and reading from storage in a variety of formats.
+#. Users should be able to define new persistence formats.
+#. Users should be able to specify where a DataSet is written.
```

Comment:
I would like to suggest some helper functions which should be possible (though I realize that this may not be for the DataSet itself. 
- It should be possible to load (parameter) settings from a DataSet onto the currently active instruments 
- It should be possible to easily compare parameter settings between different datasets and between datasets and the active environment.